sat/cli/bootprep/input/image.py

```diff
@@ -479,3 +493,124 @@ def completed_successfully(self):
         return self.image_create_success and self.image_configure_success
 
     begin = begin_image_create
+
+
+class IMSInputImage(BaseInputImage):
+    """An input image that specifies an explicit IMS image or recipe as a starting point."""
+
+    @property
+    def ims_data(self):
+        """dict: the data that defines the base IMS image or recipe"""
+        return self.image_data['ims']
+
+    @property
+    def base_is_recipe(self):
+        """bool: whether the starting point is a recipe in IMS or an image"""
+        # is_recipe is a required property
+        return self.ims_data['is_recipe']
+
+    @property
+    def base_description(self):
+        """str: a human-readable description of the base we are starting from"""
+        for param in ('id', 'name'):
+            if param in self.ims_data:
+                return f'{self.base_resource_type} with {param}={self.ims_data[param]}'
+
+        # This shouldn't happen since schema requires 'id' or 'name'
+        return self.base_resource_type
+
+    @cached_property
+    def ims_base(self):
+        """dict: the data for the base IMS recipe or image to build and/or customize"""
+        resource_type = self.base_resource_type
+        try:
+            matching_base_resources = self.ims_client.get_matching_resources(
+                resource_type, resource_id=self.ims_data.get('id'),
+                name=self.ims_data.get('name'))
+        except APIError as err:
+            raise ImageCreateError(err)
+
+        if len(matching_base_resources) == 0:
+            raise ImageCreateError(f'Found no matches for {self.base_description}')
+        elif len(matching_base_resources) > 1:
+            raise ImageCreateError(f'Found {len(matching_base_resources)} matches for {self.base_description}. '
+                                   f'This {self.base_resource_type} must be specified by "id".')
+
+        return matching_base_resources[0]
+
+
+class ProductInputImage(BaseInputImage):
+    """An input image that specifies a product-provided IMS image or recipe as a starting point."""
+
+    @property
+    def base_is_recipe(self):
+        """bool: whether the starting point is a recipe in IMS or an image"""
+        # is_recipe is a required property
+        return self.image_data['product']['is_recipe']
+
+    @property
+    def product_name(self):
+        """str: the product name"""
+        # name is a required property
+        return self.image_data['product']['name']
+
+    @property
+    @jinja_rendered
+    def product_version(self):
+        """str: the product version"""
+        # version is not required and should default to the latest version
+        return self.image_data['product'].get('version')
```

Comment:
This looks to me as though this will return `None` if `'version'` is missing from `image_data`, not the latest version. I may be missing a level of indirection here, though.

Reply:
Right, the two places this `product_version` property is used would be in the `base_description` property and in the `installed_product` cached property. The `base_description` property has logic to detect whether `self.product_version` is None. The `installed_product` cached property calls [this function in cray-product-catalog](https://github.com/Cray-HPE/cray-product-catalog/blob/a1cc1f6233616d6d45c9a3f2dfb428a8ccf02532/cray_product_catalog/query.py#L133-L170), which treats `version=None` as the latest version. I will clarify this comment to something like this:

```
# version is not required. base_description and installed_product treat None as latest
```